README.md
# generate-nft-metadata

Simple toolset to generate metadata.

## Support
Like this tool? Want to support me? Please donate some crypto to me (so I can support my 4 wives and 11 kids and 3 dogs) 

** ETH : 0xb0e73af58a9fdece76ba74a0cfb09265ae7e45d0

** Polygon : 0x70a0D3c75853f706B17970727A25113a63bCAf1f

## Art Engine
FYI , I have forked the HashLips Art Engine and made some extra features on it

https://github.com/fransyozef/hashlips_art_engine

Go check it out, maybe it can help you also.


## Index

- Setup
    - [Download](#download)
    - [Window users](#windows-users--use-gitbash-terminal-for-windows)
    - [NodeJS](#nodejs-version)
    - [Install](#install)
    - [Config.js](#config-file)
- The tools
    - [Tool 1 : Generate json file from images](#tool-1---generate-json-file-from-images)
    - [Tool 2 : Generate a collection with only 1 media](#tool-2---generate-a-collection-with-only-1-media)
    - [Tool 3 : Generate json files from the _metadata.json](#tool-3---generate-json-files-from-the-_metadatajson)
    - [Tool 4 : Update the base uri](#tool-4---update-the-base-uri)
    - [Tool 5 : Reverse generate images from _metadata.json and layers](#tool-5---reverse-generate-images-from-_metadatajson-and-layers)
    - [Tool 6 : Cleanup metadata](#tool-6---cleanup-metadata)
    - [Tool 7 : Generate _metadata.json](#tool-7---create-_metadatajson-from-seperate-json-files)



## Download 

You can download the latest release at https://github.com/fransyozef/generate-nft-metadata/releases

or you can clone this on your machine :

```sh
git clone https://github.com/fransyozef/generate-nft-metadata.git
```

## Windows users !!!! Use GitBash terminal for Windows
Some commands need some unix commands. The Windows terminal don't support them. In order to make it work you could use Git Bash as your terminal (as a replacement of PowerShell). Download it at https://gitforwindows.org/.

So use Git Bash instead of Windows PowerShell!

Without this, the scripts won't work!

## NodeJS version
This tool was tested with nodejs version v14.19.1. You need to download and install it from https://nodejs.org/en/download/.

## Install
Open your terminal and navigate to the project root folder and type :

```sh
npm install
```

## Config file

The config file is located in `/src/config.js`. This is where you configure the tools. 


## Tool #1 - Generate json file from images

Let's say you have some images, but don't have any metadata json files.

Put your images in the `/assets/tool1/` folder.

Next you can setup the default metadata that will be used in your metadata json. 
You can find this in the `config.js` file.

```
    namePrefix: 'Your Collection',
    description: 'Remember to replace this description',
    baseUri: "ipfs://NewUriToReplace/",
```


You can configure the attributes and extra metadata in the `config.js` file :

```
    extraMetadata : {
        author : "Fransjo Leihitu"
    },
    extraAttributes : [
        {
            "trait_type": "color",
            "value": "black"
        }
    ],
```

then run the command 

```sh
npm run tool1
```

this will generate the json files in `/build/json` and rename & copy your source images files to `/build/images/` folder.

So it will look something like this

```
{
  "name": "Your Collection #1",
  "description": "Remember to replace this description",
  "image": "ipfs://NewUriToReplace/1.png",
  "dna": "55d323fe03e5f28360dbf1b75f1a87ef09d0b6a2",
  "edition": 1,
  "date": 1651319815313,
  "compiler": "Metadata generator NFT by fransyozef",
  "attributes": [
    {
      "trait_type": "color",
      "value": "black"
    }
  ],
  "author": "Fransjo Leihitu"
}
````


## Tool #2 - Generate a collection with only 1 media

If you want to generate metadata for your collection witht only 1 media (png,mp4) , you can use the command

```sh
npm run generate:fixedMedia
```

In the config file you can set the filename and the editions to generate :

```
    fixedMedia : {
        filename : 'test.png',
        editions : 10
    }
```

Don't forget to update the following in your `config.js`

```
    namePrefix: 'Your Collection',
    description: 'Remember to replace this description',
    baseUri: "ipfs://NewUriToReplace/",
```

This will generate 10 metadata with all pointing to the same filename in the folder `/build/json/`

For example:

```
{
  "name": "Your Collection #1",
  "description": "Remember to replace this description",
  "image": "ipfs://NewUriToReplace/test.png",
  "dna": "f24ec0656aa4e8706480d46b7655fd07e9d293fe",
  "edition": 1,
  "date": 1651600070254,
  "compiler": "Metadata generator NFT by fransyozef",
  "attributes": [
    {
      "trait_type": "color",
      "value": "black"
    }
  ],
  "author": "Fransjo Leihitu"
}
```


## Tool #3 - Generate json files from the _metadata.json

If you want to generate json files that are stored in the _metadata.json, use the command

```sh
npm run generate:fromMetadata
```

Put the `_metadata.json` in the `/assets/tool3/` folder. The json files will be exported in the `/build/json/` folder.

So for example :

```
[
    {
        "name": "Test #1",
        "description": "My description",
        "image": "ipfs://NewUriToReplace/1.png",
        "dna": "53253828be96096cb1a321e5a3bce241db6c067f",
        "edition": 1,
        "date": 1651315338049,
        "author": "Fransjo Leihitu aka The Hangry Coder aka XperimentalConcept aka fransYOzef aka ThaLyric",
        "attributes": [],
        "compiler": "HashLips Art Engine , forked by fransyozef"
    },
    {
        "name": "Test #2",
        "description": "My description",
        "image": "ipfs://NewUriToReplace/2.png",
        "dna": "a9ab967c5e138cfb743a0d477aa4f827672eac99",
        "edition": 2,
        "date": 1651315339045,
        "author": "Fransjo Leihitu aka The Hangry Coder aka XperimentalConcept aka fransYOzef aka ThaLyric",
        "attributes": [],
        "compiler": "HashLips Art Engine , forked by fransyozef"
    }
]
```

will generate 2 json metadata files in `/build/json/`.


## Tool #4 - Update the base uri

Once you have uploaded all your images to Pinata / IPFS , you need to update the baseUri in the json files.

Open the config file in `src/config.js`  and change the `baseUri` value. 

Put your json files in the `/assets/tool4/` folder.

Next, in your terminal type

```sh
npm run updateBaseUri
```

## Tool #5 - Reverse generate images from _metadata.json and layers

This tool will generate your images by crossreferencing the `/assets/tool5/_metadata.json` and the source image(s) in `/assets/tool5/layers/`.

The images will be outputtted in `/build/images/`.

First set the correct width and height output in the `config.js`

```
    format: {
        width: 512,
        height: 512,
        smoothing: false,
    }
```

The command to run this tool :

```sh
npm run generate:fromMetadataJsonAndLayers
```

## Tool #6 - Cleanup metadata

The HashLip generator creates metadata for example

```json
{
  "name": "Your Collection #1",
  "description": "Remember to replace this description",
  "image": "ipfs://NewUriToReplace/1.png",
  "dna": "ffca094d38469d558fd8a442b847802d3ef3b42e",
  "edition": 1,
  "date": 1651601900137,
  "compiler": "Metadata generator NFT by fransyozef",
  "attributes": [
    {
      "trait_type": "color",
      "value": "black"
    }
  ],
  "author": "Fransjo Leihitu"
}
```

You don't actually need all the keys. 

Put all your metadata in the folder `/assets/tool6/` .

In `config.js` you can configure which keys you don't want :

```js
    removeMetaData : [
        "edition",
        "dna",
        "date",
        "compiler",
        "author"
    ]
```

Then run the command in your terminal

```sh
npm run cleanMetadata
```

The new metadata will be in `/build/json/`.


## Tool #7 - Create _metadata.json from seperate json files.

This tool will create the _metadata.json from the seperate generated json files.
Put your json files in `/assets/tool7/` folder and run the command

```sh
npm run generate:combinedMetadataJson
```

The `_metadata.json` will be generated in `/build/json/` folder

changelog.md
# Changelog

## 1.3.2
- Fixed Tool6 build failed when the /build folder is not there

## 1.3.1
- Added removeMetaData list for tool 6
- Bug fix tool 6: reading from wrong path
- Update readme

## 1.3.0
- Added tool #7 : Generate _metadata.json
- Update readme

## 1.2.0
- Refactored to have all input in their respective tools folder in the `/assets/` folder.
- Update readme

## 1.1.3
- Fixed typo for `generate:fixedMedia`

## 1.1.2
- Added reverse image generation from _metadata.json and layers images

## 1.1.1
- Bug fixing
- update Readme
- Added dummy attributes

## 1.1.0
- Renamed files, update README.md

## 1.0.6
- Added generate json files from _metadata.json

## 1.0.5
- Added generate with 1 fixed media

## 1.0.4
- Made `extraAttributes` optional in the configfile

## 1.0.3
- Update package.json
- Minor update

## 1.0.2
- Extra check on baseUri has a / at the end

## 1.0.1
- code cleanup
- readme cleanup

## 1.0.0

- First contact

src/application.js
const basePath = process.cwd();
const path = require('path');
const fs1 = require("fs");
const { promises: fs } = require("fs");
const mime = require('mime-types')
const sha1 = require(`${basePath}/node_modules/sha1`);
const { createCanvas, Image } = require(`${basePath}/node_modules/canvas`);
const { parse } = require('csv-parse');
const {
    ASSETS_PATH,
    BUILD_PATH,
    IMAGES_PATH,
    JSON_PATH,
    LAYERS_PATH
} = require(`${basePath}/src/constants.js`);
const {
    PROJECT_CONFIG
} = require(`${basePath}/src/config.js`);

const metadataList = [];
const filesInput = [];
const generatedJsonList = [];

let canvas, ctx;
const initCanvas = () => {
    canvas = createCanvas(PROJECT_CONFIG.format.width, PROJECT_CONFIG.format.height);
    ctx = canvas.getContext("2d");
    imageSmoothingEnabled = PROJECT_CONFIG.format.smoothing;
}

const buildSetup = () => {
    if (fs1.existsSync(BUILD_PATH)) {
        fs1.rmdirSync(BUILD_PATH, { recursive: true });
    }
    fs1.mkdirSync(BUILD_PATH);
    fs1.mkdirSync(JSON_PATH);
    fs1.mkdirSync(IMAGES_PATH);
};

const copyImage = async (sourceFile, destFile) => {
    console.log(sourceFile, destFile);
    // const sourceFile = `${ASSETS_PATH}${_prePath}${filesInput[index]}`;
    // const destFile = `${IMAGES_PATH}${index + 1}.png`;
    try {
        await fs.copyFile(sourceFile, destFile);
        const result = await fs.readFile(destFile);
        // if(result) console.log(`[${index}] copied ${sourceFile} to ${destFile}`)
    } catch (e) {
        console.log(e);
    }
}

const buildJson = (index, _media, _jsonFilename) => {
    const dna = '';
    const _attributes = [];
    let attributes = [];
    if (PROJECT_CONFIG.extraAttributes) {
        attributes = [..._attributes, ...PROJECT_CONFIG.extraAttributes];
    } else {
        attributes = _attributes;
    }
    let baseUri = PROJECT_CONFIG.baseUri;
    const baseUriCheck = baseUri.charAt(baseUri.length - 1);
    baseUri = baseUriCheck === '/' ? baseUri : `${baseUri}/`;
    const media = _media ? `${baseUri}${_media}` : `${baseUri}${index + 1}.png`;
    const tempData = {
        "name": `${PROJECT_CONFIG.namePrefix} #${index + 1}`,
        "description": `${PROJECT_CONFIG.description}`,
        "image": `${media}`,
        "dna": sha1(`${PROJECT_CONFIG.namePrefix} #${index + 1} ${Date.now()}`),
        "edition": index + 1,
        "date": Date.now(),
        "compiler": "Metadata generator NFT by fransyozef",
        "attributes": attributes
    }
    const payload = { ...tempData, ...PROJECT_CONFIG.extraMetadata };
    saveJson(payload, _jsonFilename);
    metadataList.push(payload);
}

const saveJson = (payload, _filename) => {
    fs1.writeFileSync(
        _filename,
        JSON.stringify(payload, null, 2)
    );
}

const saveAllJson = (_metadataList) => {
    const __metadataList = _metadataList ? _metadataList : metadataList;
    fs1.writeFileSync(
        `${JSON_PATH}_metadata.json`,
        JSON.stringify(__metadataList, null, 2)
    );
}
const getSourceImages = async () => {
    const tool1Path = `${ASSETS_PATH}tool1/`;
    try {
        const filenames = await fs.readdir(tool1Path);
        filenames.forEach(function (file) {
            const _file = `${tool1Path}${file}`;
            const _mime = mime.lookup(_file);
            if (_mime === 'image/png') {
                filesInput.push(file);
                console.log(`Found asset: ${file}`);
            }
        });
    } catch (e) {
        console.log(e);
    }
}

const tool1 = async () => {
    buildSetup();
    await getSourceImages();
    const totalImages = filesInput.length;
    console.log(`Generating ${totalImages} pictures and metadata`);
    for (let i = 0; i < totalImages; i++) {
        const sourceFile = `${ASSETS_PATH}tool1/${filesInput[i]}`;
        const destFile = `${IMAGES_PATH}${i + 1}.png`;
        await copyImage(sourceFile, destFile);
        buildJson(i, null, `${JSON_PATH}${i + 1}.json`);
        saveAllJson();
    }
    showSupport();
}

const getGeneratedJson = async (_path) => {
    const generatedJsonList = [];
    try {
        const filenames = await fs.readdir(_path);
        filenames.forEach(function (file) {
            const _file = `${_path}${file}`;
            const _mime = mime.lookup(_file);
            if (_mime === 'application/json' && file !== '_metadata.json') {
                generatedJsonList.push(`${_path}${file}`);
                console.log(`Found json: ${file}`);
            }
        });
        return generatedJsonList;
    } catch (e) {
        console.log(e);
        return null;
    }
}

const cleanupMetadata = async (_files) => {
    _files.forEach((file) => {
        const _exportFilename = getFilename(file);
        if (_exportFilename) {
            const exportFilename = `${JSON_PATH}${_exportFilename}`;
            let rawdata = fs1.readFileSync(file);
            let data = JSON.parse(rawdata);
            PROJECT_CONFIG.removeMetaData.forEach((metadata) => {
                if (data[metadata]) {
                    delete data[metadata];
                }
            });
            fs1.writeFileSync(
                exportFilename,
                JSON.stringify(data, null, 2)
            );
            console.log(`Wrote new metadata in ${exportFilename}`);
        } else {
            console.log('[ERROR] Could not get filename!');
        }
    });
}

const tool6 = async () => {
    buildSetup();
    const files = await getGeneratedJson(`${ASSETS_PATH}tool6/`);
    cleanupMetadata(files);
    showSupport();
}

const showSupport = () => {
    console.log('');
    console.log('Done!');
    console.log('');
    console.log(`
        Like this tool? Want to support me? Please donate some crypto to me (so I can support my 4 wives and 11 kids and 3 dogs) 

        ** ETH : 0xb0e73af58a9fdece76ba74a0cfb09265ae7e45d0
        
        ** Polygon : 0x70a0D3c75853f706B17970727A25113a63bCAf1f
    `);
}

const updateImageLocationInJson = async (files) => {
    files.forEach((file) => {
        let rawdata = fs1.readFileSync(file);
        let data = JSON.parse(rawdata);
        const _imageSplit = data.image.split('/');
        const _filename = _imageSplit.pop();
        let baseUri = PROJECT_CONFIG.baseUri;
        const baseUriCheck = baseUri.charAt(baseUri.length - 1);
        baseUri = baseUriCheck === '/' ? baseUri : `${baseUri}/`;
        data.image = `${baseUri}${_filename}`;
        const exportfileName = getFilename(file);
        if (exportfileName) {
            saveJson(data, `${JSON_PATH}${exportfileName}`);
        } else {
            console.log('[ERROR] Could not create export filename!');
        }
    });
}

const getFilename = (_filename) => {
    if (!_filename) return null;
    try {
        const parts = _filename.split('/');
        const filename = parts.pop();
        return filename;
    } catch (e) {
        return _filename;
    }
}

const tool4 = async () => {
    buildSetup();
    const files = await getGeneratedJson(`${ASSETS_PATH}tool4/`);
    if (files && files.length > 0) {
        updateImageLocationInJson(files);
    } else {
        console.log('[ERROR] No files found!')
    }
    showSupport();
}

const tool2 = () => {
    buildSetup();
    const totalMetadata = PROJECT_CONFIG.fixedMedia.editions;
    console.log(`Generating ${totalMetadata} metadata with fixed media : ${PROJECT_CONFIG.fixedMedia.filename}`);
    for (let i = 0; i < totalMetadata; i++) {
        buildJson(i, PROJECT_CONFIG.fixedMedia.filename, `${JSON_PATH}${i + 1}.json`);
        saveAllJson();
    }
    showSupport();
}

const tool3 = () => {
    buildSetup();
    const metadataJson = `${ASSETS_PATH}tool3/_metadata.json`;
    console.log(`Reading ${metadataJson}`);
    let rawdata = fs1.readFileSync(metadataJson);
    let data = JSON.parse(rawdata);
    if (data && data.length > 0) {
        data.forEach((element) => {
            const _filename = `${element.edition}.json`;
            console.log(`Exporting to ${_filename}`);
            const filename = `${JSON_PATH}${_filename}`;
            saveJson(element, filename);
        });
    }
    showSupport();
}

const tool5 = async () => {
    buildSetup();
    initCanvas();
    const metadataJson = `${ASSETS_PATH}tool5/_metadata.json`;
    console.log(`Reading ${metadataJson}`);
    console.log('');
    let rawdata = fs1.readFileSync(metadataJson);
    let data = JSON.parse(rawdata);
    let fileStack = [];
    if (data && data.length > 0) {
        for (x = 0; x < data.length; x++) {
            const element = data[x];
            if (element?.attributes) {
                fileStack = [];
                for (i = 0; i < element.attributes.length; i++) {
                    const attribute = element.attributes[i];
                    let layer = `${ASSETS_PATH}tool5/layers/${attribute.trait_type}/`;
                    let filename = attribute.value;
                    const filenames = await fs.readdir(layer);
                    filenames.forEach(function (file) {
                        let _file = file.replace('.png', '');
                        let fileParts = _file.split('#');
                        if (fileParts[0] === filename) {
                            fileStack.push(`${layer}${file}`);
                        }
                    });
                }
                if (fileStack.length > 0) {
                    for (i = 0; i < fileStack.length; i++) {
                        const image = fileStack[i];
                        const img = new Image();
                        img.onload = () => ctx.drawImage(img, 0, 0);
                        img.onerror = err => { throw err }
                        img.src = image;
                    }
                    const filename = `${IMAGES_PATH}${element.edition}.png`;
                    fs1.writeFileSync(
                        filename,
                        canvas.toBuffer("image/png")
                    );
                    console.log(`png saved at ${filename}`);
                    console.log(`metadata saved`);
                    console.log(` `);
                    const exportJson = `${JSON_PATH}${element.edition}.json`;
                    saveJson(element, exportJson);
                }

                let filename = `${JSON_PATH}_metadata.json`;
                fs1.writeFileSync(
                    filename,
                    JSON.stringify(data, null, 2)
                );

                initCanvas();
            }
        }
    }
    showSupport();
}

const tool7 = async () => {
    buildSetup();
    const jsonFilesPath = `${ASSETS_PATH}tool7/`;
    const filenames = await fs.readdir(jsonFilesPath);
    let payload = [];
    filenames.forEach(function (file) {
        if (file !== '_metadata.json') {
            const _mime = mime.lookup(file);
            if (_mime === 'application/json') {
                console.log(`Found json file: ${file}`);
                let _file = `${jsonFilesPath}${file}`;
                let rawdata = fs1.readFileSync(_file);
                let data = JSON.parse(rawdata);
                payload.push(data);
            }
        }
    });

    console.log(' ');
    if (payload.length > 0) {
        console.log('Ready to generate _metadata.json');
        const metadataJson = `${JSON_PATH}_metadata.json`;
        fs1.writeFileSync(
            metadataJson,
            JSON.stringify(payload, null, 2)
        );
        console.log(`Check in ${JSON_PATH}`);
    } else {
        console.log('No json files where found!!!!');
    }

    showSupport();
}

const tool8 = () => {
    buildSetup();
    const importCSV = `${ASSETS_PATH}tool8/import.csv`;
    const csvData = [];
    fs1.createReadStream(importCSV)
        .pipe(parse({ delimiter: '|' }))
        .on('data', function (csvrow) {
            csvData.push(csvrow);
        })
        .on('end', function () {
            if (csvData.length > 0) {
                const columnNames = csvData.shift();
                columnNames.shift();
                if (csvData.length > 0) {
                    for (let q = 0; q < csvData.length; q++) {
                        const line = csvData[q];
                        const attributes = [];
                        const fileName = line.shift();
                        const counterParts = fileName.split('.json');
                        const counter = counterParts[0];
                        for (let i = 0; i < line.length; i++) {
                            if (line[i] != '') {
                                const attribute = {
                                    "trait_type": `${columnNames[i]}`,
                                    "value": `${line[i]}`
                                };
                                attributes.push(attribute);
                            }
                        }
                        PROJECT_CONFIG.extraAttributes = attributes;
                        buildJson(counter - 1, null, `${JSON_PATH}${fileName}`);
                    }
                } else {
                    console.log('[ERROR] You have headers but no any CSV data!!!!');
                }
            } else {
                console.log('[ERROR] Could not find any CSV data!!!!');
            }
            showSupport();
        });
}

module.exports = {
    tool1,
    tool2,
    tool3,
    tool4,
    tool5,
    tool7,
    tool6,
    tool8,
    showSupport,
};
